File: iosApp/iosApp/Pages/StopDetails/TripStopRow.swift
//
//  TripStopRow.swift
//  iosApp
//
//  Created by esimon on 12/4/24.
//  Copyright © 2024 MBTA. All rights reserved.
//

import Shared
import SwiftUI

struct TripStopRow: View {
    var stop: TripDetailsStopList.Entry
    var trip: Trip
    var now: EasternTimeInstant
    var onTapLink: (TripDetailsStopList.Entry) -> Void
    var onOpenAlertDetails: (Shared.Alert) -> Void
    var route: Route
    var routeAccents: TripRouteAccents
    var alertSummaries: [String: AlertSummary?]
    var showDownstreamAlert: Bool = false
    var targeted: Bool = false
    var firstStop: Bool = false
    var lastStop: Bool = false
    var background: Color? = nil

    var activeElevatorAlerts: [Shared.Alert] {
        stop.activeElevatorAlerts(now: now)
    }

    var disruption: UpcomingFormat.Disruption? {
        if let disruption = stop.disruption, disruption.alert.hasNoThroughService, showDownstreamAlert {
            disruption
        } else {
            nil
        }
    }

    var stickConnections: [RouteBranchSegment.StickConnection] {
        RouteBranchSegment.StickConnection.companion.forward(
            stopBefore: firstStop ? nil : "",
            stop: stop.stop.id,
            stopAfter: lastStop ? nil : "",
            lane: .center
        )
        .compactMap { $0 }
    }

    var body: some View {
        StopListRow(
            stop: stop.stop,
            stopLane: .center,
            stickConnections: stickConnections,
            onClick: { onTapLink(stop) },
            routeAccents: routeAccents,
            stopListContext: .trip,
            activeElevatorAlerts: activeElevatorAlerts.count,
            alertSummaries: alertSummaries,
            background: background,
            connectingRoutes: stop.routes,
            disruption: disruption,
            getAlertState: { fromStop, _ in
                if fromStop == stop.stop.id, showDownstreamAlert, disruption?.alert.effect == .shuttle {
                    .shuttle
                } else {
                    .normal
                }
            },
            stopPlacement: .init(isFirst: firstStop, isLast: lastStop),
            onOpenAlertDetails: onOpenAlertDetails,
            targeted: targeted,
            trackNumber: stop.trackNumber,
            descriptor: { EmptyView() },
            rightSideContent: rightSideContent
        )
    }

    @ViewBuilder func rightSideContent() -> some View {
        if let upcomingTripViewState {
            UpcomingTripView(
                prediction: upcomingTripViewState,
                routeType: routeAccents.type,
                hideRealtimeIndicators: true,
                maxTextAlpha: 0.6
            ).foregroundStyle(Color.text)
        }
    }

    var upcomingTripViewState: UpcomingTripView.State? {
        guard let formatted = stop.format(trip: trip, now: now, route: route) else { return nil }
        switch onEnum(of: formatted) {
        case let .disruption(disruption): return .disruption(
                .init(alert: disruption.alert),
                iconName: disruption.iconName
            )
        case let .some(some): return .some(some.trips[0].format)
        default: return nil
        }
    }
}

#Preview {
    let objects = ObjectCollectionBuilder()
    let trip = objects.trip { _ in }
    let now = EasternTimeInstant.now()
    let red = objects.route { route in
        route.color = "DA291C"
        route.longName = "Red Line"
        route.textColor = "FFFFFF"
        route.type = .heavyRail
    }
    VStack(spacing: 0) {
        TripStopRow(
            stop: .init(
                stop: objects.stop { stop in
                    stop.name = "Charles/MGH"
                    stop.wheelchairBoarding = .accessible
                },
                stopSequence: 10,
                disruption: nil,
                schedule: nil,
                prediction: objects.prediction { $0.status = "Stopped 5 stops away" },
                vehicle: nil,
                routes: [
                    red,
                    objects.route { route in
                        route.longName = "Green Line"
                        route.color = "00843D"
                        route.textColor = "FFFFFF"
                    },
                ]
            ),
            trip: trip,
            now: now,
            onTapLink: { _ in },
            onOpenAlertDetails: { _ in },
            route: red,
            routeAccents: .init(route: red),
            alertSummaries: [:]
        )
        TripStopRow(
            stop: .init(
                stop: objects.stop { $0.name = "Park Street" },
                stopSequence: 10,
                disruption: nil,
                schedule: nil,
                prediction: objects.prediction { $0.departureTime = now.plus(minutes: 5) },
                vehicle: nil,
                routes: [
                    red,
                    objects.route { route in
                        route.longName = "Green Line"
                        route.color = "00843D"
                        route.textColor = "FFFFFF"
                    },
                ]
            ),
            trip: trip,
            now: now,
            onTapLink: { _ in },
            onOpenAlertDetails: { _ in },
            route: red,
            routeAccents: .init(route: red),
            alertSummaries: [:]
        )
        TripStopRow(
            stop: .init(
                stop: objects.stop { $0.name = "South Station" },
                stopSequence: 10,
                disruption: nil,
                schedule: nil,
                prediction: objects.prediction { $0.departureTime = now.plus(minutes: 5) },
                predictionStop: objects.stop { $0.platformCode = "1" },
                vehicle: nil,
                routes: [],
                elevatorAlerts: [
                    objects.alert {
                        $0.activePeriod(
                            start: now.minus(minutes: 20),
                            end: now.plus(minutes: 20)
                        )
                    },
                ]
            ),
            trip: trip,
            now: now,
            onTapLink: { _ in },
            onOpenAlertDetails: { _ in },
            route: red,
            routeAccents: .init(route: red),
            alertSummaries: [:]
        )
    }
    .withFixedSettings([.stationAccessibility: true])
    .font(Typography.body)
    .background(Color.fill3)
}

#Preview("Disruptions") {
    let objects = ObjectCollectionBuilder()
    let trip = objects.trip { _ in }
    let now = EasternTimeInstant.now()
    let red = objects.route { route in
        route.color = "DA291C"
        route.longName = "Red Line"
        route.textColor = "FFFFFF"
        route.type = .heavyRail
    }
    ZStack {
        Color.fill3.padding(6)
        VStack(spacing: 0) {
            TripStopRow(
                stop:
                .init(
                    stop: objects.stop { $0.name = "Charles/MGH" },
                    stopSequence: 10,
                    disruption: .init(
                        alert: objects.alert { $0.effect = .stopClosure },
                        mapStopRoute: .red
                    ),
                    schedule: nil,
                    prediction: objects.prediction { $0.status = "Stopped 5 stops away" },
                    vehicle: nil,
                    routes: [
                        red,
                        objects.route { route in
                            route.longName = "Green Line"
                            route.color = "00843D"
                            route.textColor = "FFFFFF"
                        },
                    ]
                ),
                trip: trip,
                now: now,
                onTapLink: { _ in },
                onOpenAlertDetails: { _ in },
                route: red,
                routeAccents: .init(route: red),
                alertSummaries: [:]
            )
            TripStopRow(
                stop:
                .init(
                    stop: objects.stop { $0.name = "Park Street" },
                    stopSequence: 10,
                    disruption: nil,
                    schedule: nil,
                    prediction: objects.prediction { $0.departureTime = now.plus(minutes: 5) },
                    vehicle: nil,
                    routes: [
                        red,
                        objects.route { route in
                            route.longName = "Green Line"
                            route.color = "00843D"
                            route.textColor = "FFFFFF"
                        },
                    ]
                ),
                trip: trip,
                now: now,
                onTapLink: { _ in },
                onOpenAlertDetails: { _ in },
                route: red,
                routeAccents: .init(route: red),
                alertSummaries: [:]
            )
            TripStopRow(
                stop:
                .init(
                    stop: objects.stop { $0.name = "South Station" },
                    stopSequence: 10,
                    disruption: .init(alert: objects.alert { $0.effect = .shuttle },
                                      mapStopRoute: .red),
                    schedule: nil,
                    prediction: objects.prediction { $0.departureTime = now.plus(minutes: 5) },
                    predictionStop: objects.stop { $0.platformCode = "1" },
                    vehicle: nil,
                    routes: [],
                    elevatorAlerts: []
                ),
                trip: trip,
                now: now,
                onTapLink: { _ in },
                onOpenAlertDetails: { _ in },
                route: red,
                routeAccents: .init(route: red),
                alertSummaries: [:],
                showDownstreamAlert: true
            )
        }
    }
    .padding(6)
    .withFixedSettings([.stationAccessibility: true])
}

File: androidApp/src/main/java/com/mbta/tid/mbta_app/android/stopDetails/TripStopRow.kt
package com.mbta.tid.mbta_app.android.stopDetails

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.LocalContentColor
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.colorResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.mbta.tid.mbta_app.android.MyApplicationTheme
import com.mbta.tid.mbta_app.android.R
import com.mbta.tid.mbta_app.android.component.StopListContext
import com.mbta.tid.mbta_app.android.component.StopListRow
import com.mbta.tid.mbta_app.android.component.StopPlacement
import com.mbta.tid.mbta_app.android.component.UpcomingTripView
import com.mbta.tid.mbta_app.android.component.UpcomingTripViewState
import com.mbta.tid.mbta_app.android.util.FormattedAlert
import com.mbta.tid.mbta_app.android.util.SettingsCache
import com.mbta.tid.mbta_app.android.util.containsWrappableText
import com.mbta.tid.mbta_app.android.util.fromHex
import com.mbta.tid.mbta_app.android.util.modifiers.DestinationPredictionBalance
import com.mbta.tid.mbta_app.model.Alert
import com.mbta.tid.mbta_app.model.AlertSummary
import com.mbta.tid.mbta_app.model.MapStopRoute
import com.mbta.tid.mbta_app.model.ObjectCollectionBuilder
import com.mbta.tid.mbta_app.model.Route
import com.mbta.tid.mbta_app.model.RouteBranchSegment
import com.mbta.tid.mbta_app.model.RouteType
import com.mbta.tid.mbta_app.model.SegmentAlertState
import com.mbta.tid.mbta_app.model.Trip
import com.mbta.tid.mbta_app.model.TripDetailsStopList
import com.mbta.tid.mbta_app.model.UpcomingFormat
import com.mbta.tid.mbta_app.model.WheelchairBoardingStatus
import com.mbta.tid.mbta_app.repositories.MockSettingsRepository
import com.mbta.tid.mbta_app.utils.EasternTimeInstant
import kotlin.time.Duration.Companion.minutes
import org.koin.compose.KoinContext
import org.koin.dsl.koinApplication
import org.koin.dsl.module

@Composable
fun TripStopRow(
    stop: TripDetailsStopList.Entry,
    trip: Trip,
    now: EasternTimeInstant,
    onTapLink: (TripDetailsStopList.Entry) -> Unit,
    onOpenAlertDetails: (Alert) -> Unit,
    route: Route,
    routeAccents: TripRouteAccents,
    alertSummaries: Map<String, AlertSummary?>,
    modifier: Modifier = Modifier,
    showDownstreamAlert: Boolean = false,
    targeted: Boolean = false,
    firstStop: Boolean = false,
    lastStop: Boolean = false,
) {
    val activeElevatorAlerts = stop.activeElevatorAlerts(now)

    val disruption = stop.disruption?.takeIf { it.alert.hasNoThroughService && showDownstreamAlert }

    StopListRow(
        stop = stop.stop,
        stopLane = RouteBranchSegment.Lane.Center,
        stickConnections =
            RouteBranchSegment.StickConnection.forward(
                "".takeUnless { firstStop },
                stop.stop.id,
                "".takeUnless { lastStop },
                RouteBranchSegment.Lane.Center,
            ),
        onClick = { onTapLink(stop) },
        routeAccents = routeAccents,
        stopListContext = StopListContext.Trip,
        modifier = modifier,
        activeElevatorAlerts = activeElevatorAlerts.size,
        alertSummaries = alertSummaries,
        connectingRoutes = stop.routes,
        disruption = disruption,
        getAlertState = { fromStop, toStop ->
            when {
                fromStop == stop.stop.id &&
                    showDownstreamAlert &&
                    disruption?.alert?.effect == Alert.Effect.Shuttle -> SegmentAlertState.Shuttle
                else -> SegmentAlertState.Normal
            }
        },
        stopPlacement = StopPlacement(firstStop, lastStop),
        onOpenAlertDetails = onOpenAlertDetails,
        targeted = targeted,
        trackNumber = stop.trackNumber,
        rightSideContent = { rightSideModifier ->
            CompositionLocalProvider(LocalContentColor provides colorResource(R.color.text)) {
                val state = upcomingTripViewState(stop, trip, now, route)
                if (state != null) {
                    UpcomingTripView(
                        state,
                        rightSideModifier.then(
                            DestinationPredictionBalance.predictionWidth(
                                state.containsWrappableText()
                            )
                        ),
                        routeType = routeAccents.type,
                        hideRealtimeIndicators = true,
                        maxTextAlpha = 0.6f,
                    )
                }
            }
        },
    )
}

private fun upcomingTripViewState(
    stop: TripDetailsStopList.Entry,
    trip: Trip,
    now: EasternTimeInstant,
    route: Route,
): UpcomingTripViewState? {
    return when (val formatted = stop.format(trip, now, route)) {
        is UpcomingFormat.Some ->
            UpcomingTripViewState.Some(formatted.trips.singleOrNull()?.format ?: return null)
        is UpcomingFormat.Disruption ->
            UpcomingTripViewState.Disruption(
                FormattedAlert(formatted.alert),
                iconName = formatted.iconName,
            )
        else -> null
    }
}

@Preview
@Composable
private fun TripStopRowPreview() {
    val koin = koinApplication {
        modules(module { single { SettingsCache(MockSettingsRepository()) } })
    }
    val objects = ObjectCollectionBuilder("TripStopRowPreview")
    val trip = objects.trip()
    val now = EasternTimeInstant.now()
    val red =
        objects.route {
            color = "DA291C"
            longName = "Red Line"
            textColor = "FFFFFF"
            type = RouteType.HEAVY_RAIL
        }
    val redAccents = TripRouteAccents(red)
    KoinContext(koin.koin) {
        MyApplicationTheme {
            Column(Modifier.background(colorResource(R.color.fill3))) {
                TripStopRow(
                    stop =
                        TripDetailsStopList.Entry(
                            objects.stop {
                                name = "Charles/MGH"
                                wheelchairBoarding = WheelchairBoardingStatus.ACCESSIBLE
                            },
                            stopSequence = 10,
                            disruption = null,
                            schedule = null,
                            prediction = objects.prediction { status = "Stopped 5 stops away" },
                            vehicle = null,
                            routes =
                                listOf(
                                    red,
                                    objects.route {
                                        longName = "Green Line"
                                        color = "00843D"
                                        textColor = "FFFFFF"
                                    },
                                ),
                        ),
                    trip,
                    now,
                    onTapLink = {},
                    onOpenAlertDetails = {},
                    red,
                    redAccents,
                    alertSummaries = emptyMap(),
                )
                TripStopRow(
                    stop =
                        TripDetailsStopList.Entry(
                            objects.stop { name = "Park Street" },
                            stopSequence = 10,
                            disruption = null,
                            schedule = null,
                            prediction = objects.prediction { departureTime = now.plus(5.minutes) },
                            vehicle = null,
                            routes =
                                listOf(
                                    red,
                                    objects.route {
                                        longName = "Green Line"
                                        color = "00843D"
                                        textColor = "FFFFFF"
                                    },
                                ),
                        ),
                    trip,
                    now,
                    onTapLink = {},
                    onOpenAlertDetails = {},
                    red,
                    redAccents,
                    alertSummaries = emptyMap(),
                )
                TripStopRow(
                    stop =
                        TripDetailsStopList.Entry(
                            objects.stop { name = "South Station" },
                            stopSequence = 10,
                            disruption = null,
                            schedule = null,
                            prediction = objects.prediction { departureTime = now.plus(5.minutes) },
                            predictionStop = objects.stop { platformCode = "1" },
                            vehicle = null,
                            routes = emptyList(),
                            elevatorAlerts =
                                listOf(
                                    objects.alert {
                                        activePeriod(now.minus(20.minutes), now.plus(20.minutes))
                                    }
                                ),
                        ),
                    trip,
                    now,
                    onTapLink = {},
                    onOpenAlertDetails = {},
                    objects.route { type = RouteType.COMMUTER_RAIL },
                    TripRouteAccents.default.copy(
                        type = RouteType.COMMUTER_RAIL,
                        color = Color.fromHex("DA291C"),
                    ),
                    alertSummaries = emptyMap(),
                )
            }
        }
    }
}

@Preview
@Composable
private fun TripStopRowDisruptionsPreview() {
    val koin = koinApplication {
        modules(module { single { SettingsCache(MockSettingsRepository()) } })
    }
    val objects = ObjectCollectionBuilder("TripStopRowDisruptionsPreview")
    val trip = objects.trip()
    val now = EasternTimeInstant.now()
    val red =
        objects.route {
            color = "DA291C"
            longName = "Red Line"
            textColor = "FFFFFF"
            type = RouteType.HEAVY_RAIL
        }
    val redAccents = TripRouteAccents(red)
    KoinContext(koin.koin) {
        MyApplicationTheme {
            Box {
                Box(
                    Modifier.matchParentSize()
                        .padding(6.dp)
                        .background(colorResource(R.color.fill3))
                )
                Column(Modifier.padding(top = 6.dp)) {
                    TripStopRow(
                        stop =
                            TripDetailsStopList.Entry(
                                objects.stop { name = "Charles/MGH" },
                                stopSequence = 10,
                                disruption =
                                    UpcomingFormat.Disruption(
                                        objects.alert { effect = Alert.Effect.StopClosure },
                                        mapStopRoute = MapStopRoute.RED,
                                    ),
                                schedule = null,
                                prediction = objects.prediction { status = "Stopped 5 stops away" },
                                vehicle = null,
                                routes =
                                    listOf(
                                        red,
                                        objects.route {
                                            longName = "Green Line"
                                            color = "00843D"
                                            textColor = "FFFFFF"
                                        },
                                    ),
                            ),
                        trip,
                        now,
                        onTapLink = {},
                        onOpenAlertDetails = {},
                        red,
                        redAccents,
                        alertSummaries = emptyMap(),
                        showDownstreamAlert = true,
                    )
                    TripStopRow(
                        stop =
                            TripDetailsStopList.Entry(
                                objects.stop { name = "Park Street" },
                                stopSequence = 10,
                                disruption = null,
                                schedule = null,
                                prediction =
                                    objects.prediction { departureTime = now.plus(5.minutes) },
                                vehicle = null,
                                routes =
                                    listOf(
                                        red,
                                        objects.route {
                                            longName = "Green Line"
                                            color = "00843D"
                                            textColor = "FFFFFF"
                                        },
                                    ),
                            ),
                        trip,
                        now,
                        onTapLink = {},
                        onOpenAlertDetails = {},
                        red,
                        redAccents,
                        alertSummaries = emptyMap(),
                        showDownstreamAlert = true,
                    )
                    TripStopRow(
                        stop =
                            TripDetailsStopList.Entry(
                                objects.stop { name = "South Station" },
                                stopSequence = 10,
                                disruption =
                                    UpcomingFormat.Disruption(
                                        objects.alert { effect = Alert.Effect.Shuttle },
                                        MapStopRoute.RED,
                                    ),
                                schedule = null,
                                prediction =
                                    objects.prediction { departureTime = now.plus(5.minutes) },
                                predictionStop = objects.stop { platformCode = "1" },
                                vehicle = null,
                                routes = emptyList(),
                                elevatorAlerts =
                                    listOf(
                                        objects.alert {
                                            activePeriod(
                                                now.minus(20.minutes),
                                                now.plus(20.minutes),
                                            )
                                        }
                                    ),
                            ),
                        trip,
                        now,
                        onTapLink = {},
                        onOpenAlertDetails = {},
                        objects.route { type = RouteType.COMMUTER_RAIL },
                        TripRouteAccents.default.copy(
                            type = RouteType.COMMUTER_RAIL,
                            color = Color.fromHex("DA291C"),
                        ),
                        alertSummaries = emptyMap(),
                        showDownstreamAlert = true,
                    )
                }
            }
        }
    }
}
